### DISCUSSION TITLE
How do I customize alertmanager.yaml

### DISCUSSION THREAD

**Discussion by jiangxianliang007** [prometheus-operator/kube-prometheus#995] [General]
I use kubectl patch secret update alertmanager.yaml, but it does not work.

```
kubectl patch secret alertmanager-main-generated -n monitoring  -p='{"data":{"alertmanager.yaml": "Imdsb2JhbCI6CiAgInJlc29sdmVfdGltZW91dCI6ICI1bSIKICAic210cF9zbWFydGhvc3QiOiAnc210cC4xNjMuY29tOjQ2NScKICAic210"}}'
```
---
![image](https://user-images.githubusercontent.com/24754263/103221007-9628b080-495c-11eb-8de4-8b9f2540fa63.png)

I deleted alertmanager-main-generated first, then created it, but it already exists because alertmanager-alertmanager.yaml automatically creates the default alertmanager-main-generated secret
```
kubectl delete secret alertmanager-main-generated -n monitoring
kubectl create secret generic alertmanager-main-generated --from-file=alertmanager.yaml -n monitoring
secret "alertmanager-main-generated" deleted
Error from server (AlreadyExists): secrets "alertmanager-main-generated" already exists

```

How do I customize alertmanager.yaml？

**Comment by olegmayko** [comment]
if you use jsonnet to generate manifests you may have a look at this [example](https://github.com/prometheus-operator/kube-prometheus/blob/master/examples/alertmanager-config-external.jsonnet)

**Comment by jiangxianliang007** [comment]
> jsonnet
If you don't use jsonnet, is there any other way to remove the default alertmanager-main-generated?

**Comment by shinebayar-g** [comment]
I also can't find any reference of what is using `alertmanager-main` secret. AFAIK there is no reference in `manifests`. Does prometheus-operator itself expects the secret by default?

Found it: https://github.com/prometheus-operator/prometheus-operator/blob/master/Documentation/user-guides/alerting.md#manually-managed-secret

You should edit `alertmanager-main` secret. Not -generated one.

**Comment by okhuz** [comment]
I am also confused about that I am creating the manifest and in the CRD configSecret states that it defaults to alertmanater-{name} but it still is trying to mount the -generated one? What am I missing there?

**Comment by paulfantom** [comment]
`alertmanager-{name}` secret is picked up by prometheus-operator which then creates `alertmanager-{name}-generated` secret that is mounted in alertmanager pods.

**Comment by okhuz** [comment]
It was not generating from the secret that I have created actually. I needed to edit the statefulset in order to get configuration from ```alertmanager-{name}```. Now everything is working as expected. If that is the way config maps intended to work, I haven't been able to configure it that way.

**Comment by thavlik** [comment]
I'm on `release-0.7` and not sure how to get the operator to refresh the generated secret. I can modify the StatefulSet to use alertmanager-main secret (as opposed to alertmanager-main-generated), but the controller will quickly revert the changes. Seems like the secret is going stale and there's no good way to refresh it?

**Comment by novvoo** [comment]
同样问题 
1.修改
kubectl edit sts alertmanager-main -n kubesphere-monitoring-system
里面的alertmanager-main-generated  alertmanager-main-add,保存完就变回去了。
2.
alertmanager-main-generated 这个secret删除了也会自动生成，这个是哪里控制的

**Reply by paulfantom** [reply]
@jingslunt do not modify the STS as changes to STS will be reverted by the prometheus-operator. Use `Prometheus` object for this.

`alertmanager-main-generated` is generated by the prometheus-operator after validating `alertmanager-main` secret. The latter is up for user modifications. More in https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/user-guides/alerting.md#manually-managed-secret

**Reply by naviat** [reply]
I'm not clear this way, when we create secret manually, I think we still need to edit STS alertmanager because it is generated by prometheus-operator, right?
